### PR TITLE
Auto-select python and runner versions

### PR DESCRIPTION
**Pull request by joemarshall**
This is a minor update:

a) adds support for changing python version based on pyodide version once pyodide upgrades to 3.11
b) changes ubuntu-latest to ubuntu-20.04 if you try to test something using python 3.10.2 with ubuntu-latest

I know the OS version has been explicitly fixed in pytest github, but it was breaking my CI builds where I'd got ubuntu-latest pointing at this action. Once more than one version of pyodide is supported, we'll need part (a) anyway, and I think part (b) is worth having also, as people shouldn't have to know about which versions of ubuntu are correct for which pyodide version.

### CODE REVIEW

File: .github/workflows/main.yaml [hunk 1]
```diff
@@ -41,16 +41,47 @@ on:
...
+            *)
+              echo "python=3.10.2" >> $GITHUB_OUTPUT
+              echo "pythonexec=python310" >> $GITHUB_OUTPUT
+              echo "Using python 3.10.2 for pyodide ${{inputs.pyodide-version}}"
+              ;;
+          esac
+      - name: Fix runner os
+        id: fix_runner_os
+        run: |
+          if [ "${{ steps.get_python_version.outputs.python }}" = "3.10.2"] && [ "${{ inputs.os}}" = "ubuntu-latest" ]
```
Comment (ryanking13) [inline .github/workflows/main.yaml:73]: Could you also add `ubuntu-22.04`?

**Review by ryanking13 — APPROVED** [review]
Thanks, @joemarshall. Could you please add a test (maybe an entry in the test matrix), which tests the os fallback behavior?

In the future, I think we should change python version to be auto-selected from pyodide version as Python 3.11 will land in Pyodide in the near future. (or actually, the host Python version for `pytest-pyodide` doesn't need to be exactly matched with pyodide Python version, so maybe we can explore if it is okay to simply use python 3.10 (without considering patch version))

File: .github/workflows/main.yaml (hunk 1 above)
Comment (joemarshall) [reply .github/workflows/main.yaml]: Done these. 

The auto-python version choosing is already in this code - if you add a case to the case statement in main.yaml at line 58, you can set different python versions based on pyodide versions.

**Comment by ryanking13** [comment]
It looks like `macos` runner is also replaced with `ubuntu`?
(https://github.com/pyodide/pytest-pyodide/actions/runs/3540775052/jobs/5944236085#step:3:12)

Also, would you mind moving the `get_versions` job into a `test` job so that we can hide it from github CI status? For now, it is a bit verbose I think.

image

**Comment by joemarshall** [comment]
Okay, I've fixed the macos bug, missing space in a bash if statement, sorry. 

Added a changelog. 

The setting versions bit can't go inside the test job, because it has to set runner version which has to happen outside the test job. I've made it so that it has a clearer name which is the best we can do I think, until github provide an option to hide jobs.

**Comment by joemarshall** [comment]
It did some weird test fail in hypothesis which I think is just a fluke, but other than that it appears to build okay.

**Comment by ryanking13** [comment]
Thanks, @joemarshall!

> The setting versions bit can't go inside the test job, because it has to set runner version which has to happen outside the test job. I've made it so that it has a clearer name which is the best we can do I think, until github provide an option to hide jobs.

That make sense, thanks. Probably we need to experiment using 3.10.x in host Python without pinning patch version, so we don't have to downgrade os version.

> It did some weird test fail in hypothesis which I think is just a fluke, but other than that it appears to build okay.

Yes, hypothesis is somtimes flaky.

**Pull request merged**